File: packages/agentguard-langgraph-adapter/agentguard_langgraph_adapter/product_manifest.py
"""Protected local expectations and independently observed LangGraph identity."""

from __future__ import annotations

import hashlib
import json
import os
from pathlib import Path
import stat
from typing import Any, Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    PrivateAttr,
    field_validator,
    model_validator,
)

from .activation_ack import DIGEST_PATTERN, IDENTIFIER_PATTERN, ProductActivationError

PRODUCT_EVENT_TYPES = (
    "context_assembled",
    "memory_write_proposed",
    "message_send_proposed",
    "model_input_prepared",
    "model_output_produced",
    "tool_call_proposed",
    "tool_result_produced",
)
_ENFORCEMENT = {
    "context_assembled": "pre_execution_c1",
    "memory_write_proposed": "pre_execution_c3",
    "message_send_proposed": "pre_execution_c3",
    "model_input_prepared": "pre_execution_c1",
    "model_output_produced": "post_execution_isolation",
    "tool_call_proposed": "pre_execution_c3",
    "tool_result_produced": "post_execution_isolation",
}
_MAX_MANIFEST_BYTES = 128 * 1024


def canonical_sha256(value: Any) -> str:
    """Match the repository's restricted canonical JSON, without Core imports."""

    def validate(item: Any) -> None:
        if item is None or type(item) in (str, bool, int):
            return
        if type(item) is list:
            for entry in item:
                validate(entry)
            return
        if type(item) is dict and all(type(key) is str for key in item):
            for entry in item.values():
                validate(entry)
            return
        raise ProductActivationError("invalid_canonical_value")

    try:
        validate(value)
        body = json.dumps(
            value, sort_keys=True, ensure_ascii=False, separators=(",", ":")
        )
        return "sha256:" + hashlib.sha256(body.encode("utf-8")).hexdigest()
    except ProductActivationError:
        raise
    except Exception:
        raise ProductActivationError("invalid_canonical_value") from None


class _FrozenModel(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
        frozen=True,
        strict=True,
        hide_input_in_errors=True,
        revalidate_instances="always",
    )


class LangGraphEventCapabilityV2(_FrozenModel):
    event_type: str
    supported: bool
    active: bool
    enforcement: str
    residual_boundaries: tuple[str, ...]

    @field_validator("residual_boundaries", mode="before")
    @classmethod
    def _freeze_residual(cls, value: Any) -> Any:
        return tuple(value) if isinstance(value, list) else value


class LangGraphCapabilityReportV2(_FrozenModel):
    schema_version: Literal["2.0"]
    runtime: Literal["langgraph"]
    agent_id: str = Field(min_length=1, max_length=128, pattern=IDENTIFIER_PATTERN)
    runtime_binding_id: str = Field(
        min_length=1, max_length=256, pattern=IDENTIFIER_PATTERN
    )
    profile_id: Literal["agentguard-langgraph-v2"]
    supported: bool
    active: bool
    c0_registration: bool
    c1_pre_execution_interception: bool
    c2_correlation: bool
    c3_atomic_replace_and_seal: bool
    c4_outcome_receipts: bool
    events: tuple[LangGraphEventCapabilityV2, ...]
    residual_boundaries: tuple[str, ...]
    report_digest: str = Field(pattern=DIGEST_PATTERN)

    @field_validator("events", "residual_boundaries", mode="before")
    @classmethod
    def _freeze_arrays(cls, value: Any) -> Any:
        return tuple(value) if isinstance(value, list) else value

    @model_validator(mode="after")
    def _validate_profile(self) -> LangGraphCapabilityReportV2:
        if tuple(event.event_type for event in self.events) != PRODUCT_EVENT_TYPES:
            raise ValueError("capability event set/order differs from frozen profile")
        if not all(
            (
                self.c0_registration,
                self.c1_pre_execution_interception,
                self.c2_correlation,
                self.c3_atomic_replace_and_seal,
                self.c4_outcome_receipts,
            )
        ):
            raise ValueError("capability levels differ from frozen LangGraph profile")
        if self.active and not self.supported:
            raise ValueError("active capability requires support")
        if self.residual_boundaries or any(
            event.residual_boundaries for event in self.events
        ):
            raise ValueError("LangGraph cannot carry residual boundaries")
        if any(
            not event.supported
            or event.active != self.active
            or event.enforcement != _ENFORCEMENT[event.event_type]
            for event in self.events
        ):
            raise ValueError("event enforcement differs from frozen LangGraph profile")
        if (
            canonical_sha256(self.model_dump(mode="json", exclude={"report_digest"}))
            != self.report_digest
        ):
            raise ValueError("capability report digest mismatch")
        return self


class ProductRuntimeObservation(_FrozenModel):
    runtime: Literal["langgraph"]
    runtime_version: str
    plugin_version: str
    loaded: bool
    enforcement_mode: Literal["enforce", "observe", "disabled"]
    adapter_artifact_digest: str = Field(pattern=DIGEST_PATTERN)
    host_inventory_digest: str = Field(pattern=DIGEST_PATTERN)
    tool_inventory_digest: str = Field(pattern=DIGEST_PATTERN)
    capability_report: LangGraphCapabilityReportV2


class ProductActivationManifest(_FrozenModel):
    schema_version: Literal["1.0"]
    runtime: Literal["langgraph"]
    runtime_version: Literal["1.2.7"]
    plugin_version: Literal["0.1.0rc1"]
    principal_id: str = Field(min_length=1, max_length=128, pattern=IDENTIFIER_PATTERN)
    agent_id: str = Field(min_length=1, max_length=128, pattern=IDENTIFIER_PATTERN)
    runtime_binding_id: str = Field(
        min_length=1, max_length=256, pattern=IDENTIFIER_PATTERN
    )
    profile_id: Literal["agentguard-langgraph-v2"]
    profile_digest: str = Field(pattern=DIGEST_PATTERN)
    activation_ref_digest: str = Field(pattern=DIGEST_PATTERN)
    adapter_artifact_digest: str = Field(pattern=DIGEST_PATTERN)
    capability_report_digest: str = Field(pattern=DIGEST_PATTERN)
    host_inventory_digest: str = Field(pattern=DIGEST_PATTERN)
    tool_inventory_digest: str = Field(pattern=DIGEST_PATTERN)

    _source_path: Path | None = PrivateAttr(default=None)
    _source_fingerprint: tuple[Any, ...] | None = PrivateAttr(default=None)

    @classmethod
    def from_file(cls, path: str | Path) -> ProductActivationManifest:
        normalized = _manifest_path(path)
        body, fingerprint = _read_protected_manifest(normalized)
        try:
            payload = json.loads(body.decode("utf-8"), object_pairs_hook=_unique_object)
            result = cls.model_validate(payload)
        except Exception:
            raise ProductActivationError("manifest_invalid") from None
        result._source_path = normalized
        result._source_fingerprint = fingerprint
        return result

    def assert_unchanged(self) -> None:
        if self._source_path is None or self._source_fingerprint is None:
            raise ProductActivationError("manifest_not_file_backed")
        try:
            _, fingerprint = _read_protected_manifest(self._source_path)
            if fingerprint != self._source_fingerprint:
                raise ProductActivationError("manifest_changed")
        except Exception:
            raise ProductActivationError("manifest_changed") from None

    def make_heartbeat(self, observation: ProductRuntimeObservation) -> dict[str, Any]:
        try:
            observed = ProductRuntimeObservation.model_validate(observation)
        except Exception:
            raise ProductActivationError("observation_invalid") from None
        report = observed.capability_report
        pairs = (
            (observed.runtime, self.runtime),
            (observed.runtime_version, self.runtime_version),
            (observed.plugin_version, self.plugin_version),
            (observed.adapter_artifact_digest, self.adapter_artifact_digest),
            (observed.host_inventory_digest, self.host_inventory_digest),
            (observed.tool_inventory_digest, self.tool_inventory_digest),
            (report.agent_id, self.agent_id),
            (report.runtime_binding_id, self.runtime_binding_id),
            (report.profile_id, self.profile_id),
            (report.report_digest, self.capability_report_digest),
        )
        if (
            not observed.loaded
            or observed.enforcement_mode != "enforce"
            or not report.supported
            or not report.active
            or any(actual != expected for actual, expected in pairs)
        ):
            raise ProductActivationError("observation_drift")
        return {
            "schema_version": "2.0",
            "status": "loaded",
            "loaded": True,
            "runtime_id": "langgraph",
            "agent_id": self.agent_id,
            "runtime_binding_id": self.runtime_binding_id,
            "profile_id": self.profile_id,
            "runtime_version": observed.runtime_version,
            "plugin_version": observed.plugin_version,
            "profile_digest": self.profile_digest,
            "adapter_artifact_digest": observed.adapter_artifact_digest,
            "reported_activation_ref_digest": self.activation_ref_digest,
            "host_inventory_digest": observed.host_inventory_digest,
            "plugin_inventory_digest": None,
            "plugin_order_inventory_digest": None,
            "tool_inventory_digest": observed.tool_inventory_digest,
            "capability_report": report.model_dump(mode="json"),
            "source": "agentguard-langgraph-adapter",
            "enforcement_mode": "enforce",
        }


def _unique_object(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    value: dict[str, Any] = {}
    for key, item in pairs:
        if key in value:
            raise ValueError("duplicate JSON field")
        value[key] = item
    return value


def _manifest_path(value: str | Path) -> Path:
    try:
        raw = os.fspath(value)
        if (
            not isinstance(raw, str)
            or not os.path.isabs(raw)
            or os.path.normpath(raw) != raw
        ):
            raise ValueError
        return Path(raw)
    except (ValueError, TypeError):
        raise ProductActivationError("manifest_invalid_path") from None


def _stat_identity(value: os.stat_result) -> tuple[int, ...]:
    return (
        value.st_dev,
        value.st_ino,
        value.st_uid,
        value.st_mode,
        value.st_nlink,
        value.st_size,
        value.st_mtime_ns,
        value.st_ctime_ns,
    )


def _read_protected_manifest(path: Path) -> tuple[bytes, tuple[Any, ...]]:
    descriptor: int | None = None
    try:
        owner = os.getuid()
        parent = path.parent.lstat()
        if (
            not stat.S_ISDIR(parent.st_mode)
            or parent.st_uid != owner
            or stat.S_IMODE(parent.st_mode) != 0o700
            or path.resolve(strict=True) != path
        ):
            raise ProductActivationError("manifest_insecure")
        descriptor = os.open(
            path, os.O_RDONLY | os.O_NOFOLLOW | os.O_CLOEXEC | os.O_NONBLOCK
        )
        before = os.fstat(descriptor)
        if (
            not stat.S_ISREG(before.st_mode)
            or before.st_uid != owner
            or stat.S_IMODE(before.st_mode) != 0o600
            or before.st_nlink != 1
        ):
            raise ProductActivationError("manifest_insecure")
        if before.st_size > _MAX_MANIFEST_BYTES:
            raise ProductActivationError("manifest_too_large")
        chunks = []
        size = 0
        while True:
            chunk = os.read(descriptor, min(65536, _MAX_MANIFEST_BYTES + 1 - size))
            if not chunk:
                break
            chunks.append(chunk)
            size += len(chunk)
            if size > _MAX_MANIFEST_BYTES:
                raise ProductActivationError("manifest_too_large")
        after = os.fstat(descriptor)
        parent_after = path.parent.lstat()
        if (
            _stat_identity(before) != _stat_identity(after)
            or _stat_identity(path.lstat()) != _stat_identity(after)
            or (parent.st_dev, parent.st_ino, parent.st_uid, parent.st_mode)
            != (
                parent_after.st_dev,
                parent_after.st_ino,
                parent_after.st_uid,
                parent_after.st_mode,
            )
            or path.resolve(strict=True) != path
        ):
            raise ProductActivationError("manifest_changed")
        body = b"".join(chunks)
        return body, (
            (parent.st_dev, parent.st_ino, parent.st_uid, parent.st_mode),
            _stat_identity(after),
            hashlib.sha256(body).hexdigest(),
        )
    except ProductActivationError:
        raise
    except Exception:
        raise ProductActivationError("manifest_unavailable") from None
    finally:
        if descriptor is not None:
            os.close(descriptor)

File: packages/agentguard-openclaw-plugin/src/runtime/product-manifest.ts
import { createHash } from "node:crypto";
import { constants, type BigIntStats } from "node:fs";
import {
  lstat,
  open,
  readFile,
  realpath,
  type FileHandle,
} from "node:fs/promises";
import { createRequire } from "node:module";
import { dirname, isAbsolute, normalize, parse, join } from "node:path";
import type { OpenClawActivationAckIdentity } from "./activation-ack.js";
import { restrictedCanonicalJson, restrictedDigest } from "./canonical.js";

const CONSTRUCTION_KEY = Symbol("product-manifest");
const MAX_BYTES = 128 * 1024;
const DIGEST = /^sha256:[0-9a-f]{64}$/u;
const IDENTIFIER = /^[A-Za-z0-9][A-Za-z0-9._:-]*$/u;
const RUNTIME_VERSION = "2026.7.1-2";
const PLUGIN_VERSION = "0.1.0-rc.1";
const PROFILE_ID = "agentguard-openclaw-v2-restricted";
const RESIDUALS = Object.freeze([
  "openclaw_has_no_authoritative_invocation_start_hook",
  "openclaw_hook_cannot_atomically_replace_and_seal_final_action",
  "openclaw_message_sending_host_exception_or_timeout_can_fail_open",
  "openclaw_non_tool_memory_write_has_no_native_pre_execution_hook",
  "openclaw_sync_persistence_hooks_cannot_await_remote_decision_or_rollback",
]);
const EVENT_PROFILE = [
  ["context_assembled", "pre_execution_c1", []],
  [
    "memory_write_proposed",
    "pre_execution_c1",
    [RESIDUALS[0], RESIDUALS[1], RESIDUALS[3]],
  ],
  [
    "message_send_proposed",
    "pre_execution_c1",
    [RESIDUALS[0], RESIDUALS[1], RESIDUALS[2]],
  ],
  ["model_input_prepared", "pre_execution_c1", []],
  ["model_output_produced", "post_execution_isolation", []],
  ["tool_call_proposed", "pre_execution_c1", [RESIDUALS[0], RESIDUALS[1]]],
  ["tool_result_produced", "post_execution_isolation", [RESIDUALS[4]]],
] as const;
const MANIFEST_FIELDS = [
  "schema_version",
  "runtime",
  "runtime_version",
  "plugin_version",
  "principal_id",
  "agent_id",
  "runtime_binding_id",
  "profile_id",
  "profile_digest",
  "activation_ref_digest",
  "adapter_artifact_digest",
  "capability_report_digest",
  "host_inventory_digest",
  "plugin_inventory_digest",
  "plugin_order_inventory_digest",
  "tool_inventory_digest",
] as const;
const OBSERVATION_FIELDS = [
  "runtime",
  "runtime_version",
  "plugin_version",
  "loaded",
  "enforcement_mode",
  "adapter_artifact_digest",
  "host_inventory_digest",
  "plugin_inventory_digest",
  "plugin_order_inventory_digest",
  "tool_inventory_digest",
  "capability_report",
] as const;
const CAPABILITY_FIELDS = [
  "schema_version",
  "runtime",
  "agent_id",
  "runtime_binding_id",
  "profile_id",
  "supported",
  "active",
  "c0_registration",
  "c1_pre_execution_interception",
  "c2_correlation",
  "c3_atomic_replace_and_seal",
  "c4_outcome_receipts",
  "events",
  "residual_boundaries",
  "report_digest",
] as const;

export class OpenClawProductActivationError extends Error {
  readonly code: string;
  constructor(code: string) {
    super(`OpenClaw Product activation unavailable: ${code}`);
    this.name = "OpenClawProductActivationError";
    this.code = code;
  }
}

export type OpenClawProductManifestData = Readonly<{
  schema_version: "1.0";
  runtime: "openclaw";
  runtime_version: "2026.7.1-2";
  plugin_version: "0.1.0-rc.1";
  principal_id: string;
  agent_id: string;
  runtime_binding_id: string;
  profile_id: "agentguard-openclaw-v2-restricted";
  profile_digest: string;
  activation_ref_digest: string;
  adapter_artifact_digest: string;
  capability_report_digest: string;
  host_inventory_digest: string;
  plugin_inventory_digest: string;
  plugin_order_inventory_digest: string;
  tool_inventory_digest: string;
}>;
export type OpenClawProductCapabilityReport = Readonly<{
  schema_version: "2.0";
  runtime: "openclaw";
  agent_id: string;
  runtime_binding_id: string;
  profile_id: "agentguard-openclaw-v2-restricted";
  supported: boolean;
  active: boolean;
  c0_registration: true;
  c1_pre_execution_interception: true;
  c2_correlation: true;
  c3_atomic_replace_and_seal: false;
  c4_outcome_receipts: true;
  events: readonly Readonly<{
    event_type: string;
    supported: boolean;
    active: boolean;
    enforcement: string;
    residual_boundaries: readonly string[];
  }>[];
  residual_boundaries: readonly string[];
  report_digest: string;
}>;
export type OpenClawProductRuntimeObservation = Readonly<{
  runtime: "openclaw";
  runtime_version: string;
  plugin_version: string;
  loaded: boolean;
  enforcement_mode: "enforce" | "observe" | "disabled";
  adapter_artifact_digest: string;
  host_inventory_digest: string;
  plugin_inventory_digest: string;
  plugin_order_inventory_digest: string;
  tool_inventory_digest: string;
  capability_report: OpenClawProductCapabilityReport;
}>;

/** Expected local identity, loaded only from a protected canonical file. */
export class OpenClawProductManifest {
  #data: OpenClawProductManifestData;
  #identity: Readonly<OpenClawActivationAckIdentity>;
  #path: string;
  #fingerprint: string;

  private constructor(
    key: symbol,
    data: OpenClawProductManifestData,
    path: string,
    fingerprint: string,
  ) {
    if (key !== CONSTRUCTION_KEY) fail("manifest_not_file_backed");
    this.#data = data;
    this.#path = path;
    this.#fingerprint = fingerprint;
    const {
      runtime_version,
      plugin_version,
      agent_id,
      runtime_binding_id,
      profile_id,
      activation_ref_digest,
      host_inventory_digest,
      plugin_inventory_digest,
      plugin_order_inventory_digest,
      tool_inventory_digest,
      capability_report_digest,
    } = data;
    this.#identity = Object.freeze({
      runtime_version,
      plugin_version,
      agent_id,
      runtime_binding_id,
      profile_id,
      activation_ref_digest,
      capability_digest: capability_report_digest,
      host_inventory_digest,
      plugin_inventory_digest,
      plugin_order_inventory_digest,
      tool_inventory_digest,
    });
    Object.freeze(this);
  }

  static async fromFile(path: string): Promise<OpenClawProductManifest> {
    if (
      typeof path !== "string" ||
      !isAbsolute(path) ||
      normalize(path) !== path
    )
      fail("manifest_invalid_path");
    const { body, fingerprint } = await readProtected(path);
    let data: OpenClawProductManifestData;
    try {
      const raw: unknown = JSON.parse(body);
      const canonical = restrictedCanonicalJson(raw);
      if (body !== canonical && body !== `${canonical}\n`)
        fail("manifest_not_canonical");
      const value = record(raw, MANIFEST_FIELDS);
      if (
        value.schema_version !== "1.0" ||
        value.runtime !== "openclaw" ||
        value.runtime_version !== RUNTIME_VERSION ||
        value.plugin_version !== PLUGIN_VERSION ||
        value.profile_id !== PROFILE_ID ||
        !identifier(value.principal_id, 256) ||
        !identifier(value.agent_id, 128) ||
        !identifier(value.runtime_binding_id, 256) ||
        MANIFEST_FIELDS.filter((field) => field.endsWith("_digest")).some(
          (field) => !digest(value[field]),
        )
      )
        fail("manifest_invalid");
      data = Object.freeze(value) as OpenClawProductManifestData;
    } catch (error) {
      if (error instanceof OpenClawProductActivationError) throw error;
      fail("manifest_invalid");
    }
    return new OpenClawProductManifest(
      CONSTRUCTION_KEY,
      data,
      path,
      fingerprint,
    );
  }

  static isManifest(value: unknown): value is OpenClawProductManifest {
    return typeof value === "object" && value !== null && #path in value;
  }
  get data(): OpenClawProductManifestData {
    return this.#data;
  }
  get expectedAckIdentity(): Readonly<OpenClawActivationAckIdentity> {
    return this.#identity;
  }

  async assertUnchanged(): Promise<void> {
    try {
      const current = await readProtected(this.#path);
      if (current.fingerprint !== this.#fingerprint) fail("manifest_changed");
    } catch {
      fail("manifest_changed");
    }
  }

  async assertInstalledVersions(): Promise<void> {
    try {
      const own = JSON.parse(
        await readFile(new URL("../../package.json", import.meta.url), "utf8"),
      );
      let cursor = dirname(
        await realpath(
          createRequire(import.meta.url).resolve(
            "openclaw/plugin-sdk/agent-harness",
          ),
        ),
      );
      let host: { name?: unknown; version?: unknown } | undefined;
      while (true) {
        try {
          const candidate = JSON.parse(
            await readFile(join(cursor, "package.json"), "utf8"),
          );
          if (candidate.name === "openclaw") {
            host = candidate;
            break;
          }
        } catch {
          /* Walk the actual resolved SDK's package ancestry. */
        }
        if (cursor === parse(cursor).root) break;
        cursor = dirname(cursor);
      }
      if (
        own.name !== "@agentguard-ai/openclaw-plugin" ||
        own.version !== PLUGIN_VERSION ||
        host?.version !== RUNTIME_VERSION
      )
        fail("version_mismatch");
    } catch (error) {
      if (error instanceof OpenClawProductActivationError) throw error;
      fail("runtime_version_unavailable");
    }
  }

  makeHeartbeat(observed: unknown): Record<string, unknown> {
    const value = readOpenClawProductRuntimeObservation(observed);
    const report = value.capability_report;
    const data = this.#data;
    if (
      !value.loaded ||
      value.enforcement_mode !== "enforce" ||
      !report.active ||
      !report.supported ||
      value.runtime_version !== data.runtime_version ||
      value.plugin_version !== data.plugin_version ||
      report.agent_id !== data.agent_id ||
      report.runtime_binding_id !== data.runtime_binding_id ||
      report.report_digest !== data.capability_report_digest ||
      [
        "adapter_artifact_digest",
        "host_inventory_digest",
        "plugin_inventory_digest",
        "plugin_order_inventory_digest",
        "tool_inventory_digest",
      ].some(
        (field) =>
          value[field as keyof typeof value] !==
          data[field as keyof typeof data],
      )
    )
      fail("observation_drift");
    return {
      schema_version: "2.0",
      status: "loaded",
      loaded: true,
      runtime_id: "openclaw",
      agent_id: data.agent_id,
      runtime_binding_id: data.runtime_binding_id,
      profile_id: data.profile_id,
      runtime_version: value.runtime_version,
      plugin_version: value.plugin_version,
      profile_digest: data.profile_digest,
      adapter_artifact_digest: value.adapter_artifact_digest,
      reported_activation_ref_digest: data.activation_ref_digest,
      host_inventory_digest: value.host_inventory_digest,
      plugin_inventory_digest: value.plugin_inventory_digest,
      plugin_order_inventory_digest: value.plugin_order_inventory_digest,
      tool_inventory_digest: value.tool_inventory_digest,
      capability_report: report,
      source: "agentguard-openclaw-plugin",
      enforcement_mode: value.enforcement_mode,
    };
  }
}

/** Copy own JSON data and validate the complete frozen restricted profile. */
export function readOpenClawProductRuntimeObservation(
  value: unknown,
): OpenClawProductRuntimeObservation {
  try {
    const raw = record(copyJson(value), OBSERVATION_FIELDS);
    if (
      raw.runtime !== "openclaw" ||
      typeof raw.runtime_version !== "string" ||
      typeof raw.plugin_version !== "string" ||
      typeof raw.loaded !== "boolean" ||
      !["enforce", "observe", "disabled"].includes(
        raw.enforcement_mode as string,
      ) ||
      OBSERVATION_FIELDS.filter((key) => key.endsWith("_digest")).some(
        (key) => !digest(raw[key]),
      )
    )
      fail("observation_invalid");
    const report = record(raw.capability_report, CAPABILITY_FIELDS);
    if (
      report.schema_version !== "2.0" ||
      report.runtime !== "openclaw" ||
      report.profile_id !== PROFILE_ID ||
      !identifier(report.agent_id, 128) ||
      !identifier(report.runtime_binding_id, 256) ||
      typeof report.active !== "boolean" ||
      typeof report.supported !== "boolean" ||
      (report.active && !report.supported) ||
      [
        "c0_registration",
        "c1_pre_execution_interception",
        "c2_correlation",
        "c4_outcome_receipts",
      ].some((field) => report[field] !== true) ||
      report.c3_atomic_replace_and_seal !== false ||
      !Array.isArray(report.events) ||
      report.events.length !== EVENT_PROFILE.length ||
      restrictedCanonicalJson(report.residual_boundaries) !==
        restrictedCanonicalJson(RESIDUALS)
    )
      fail("observation_invalid");
    for (let index = 0; index < EVENT_PROFILE.length; index += 1) {
      const event = record(report.events[index], [
        "event_type",
        "supported",
        "active",
        "enforcement",
        "residual_boundaries",
      ]);
      const [name, enforcement, residuals] = EVENT_PROFILE[index];
      if (
        event.event_type !== name ||
        event.supported !== true ||
        event.active !== report.active ||
        event.enforcement !== enforcement ||
        restrictedCanonicalJson(event.residual_boundaries) !==
          restrictedCanonicalJson(residuals)
      )
        fail("observation_invalid");
    }
    const { report_digest, ...projection } = report;
    if (
      !digest(report_digest) ||
      restrictedDigest(projection) !== report_digest
    )
      fail("observation_invalid");
    return deepFreeze(raw) as OpenClawProductRuntimeObservation;
  } catch {
    fail("observation_invalid");
  }
}

function fail(code: string): never {
  throw new OpenClawProductActivationError(code);
}
function digest(value: unknown): boolean {
  return typeof value === "string" && DIGEST.test(value);
}
function identifier(value: unknown, maximum: number): boolean {
  return (
    typeof value === "string" &&
    value.length <= maximum &&
    IDENTIFIER.test(value)
  );
}
function record(
  value: unknown,
  fields: readonly string[],
): Record<string, unknown> {
  if (
    typeof value !== "object" ||
    value === null ||
    Array.isArray(value) ||
    ![Object.prototype, null].includes(Object.getPrototypeOf(value))
  )
    fail("invalid_data");
  const keys = Reflect.ownKeys(value);
  if (
    keys.length !== fields.length ||
    keys.some((key) => typeof key !== "string" || !fields.includes(key))
  )
    fail("invalid_data");
  const result: Record<string, unknown> = {};
  for (const key of fields) {
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (!descriptor || !descriptor.enumerable || !("value" in descriptor))
      fail("invalid_data");
    result[key] = descriptor.value;
  }
  return result;
}
function copyJson(value: unknown, active = new Set<object>()): unknown {
  if (value === null || ["string", "boolean", "number"].includes(typeof value))
    return value;
  if (typeof value !== "object" || active.has(value)) fail("invalid_data");
  active.add(value);
  try {
    const array = Array.isArray(value);
    if (
      !array &&
      ![Object.prototype, null].includes(Object.getPrototypeOf(value))
    )
      fail("invalid_data");
    const result: unknown[] | Record<string, unknown> = array ? [] : {};
    const keys = Reflect.ownKeys(value);
    if (array && keys.length !== value.length + 1) fail("invalid_data");
    for (const key of keys) {
      if (array && key === "length") continue;
      if (
        typeof key !== "string" ||
        (array && !/^(?:0|[1-9][0-9]*)$/u.test(key))
      )
        fail("invalid_data");
      const descriptor = Object.getOwnPropertyDescriptor(value, key);
      if (!descriptor || !descriptor.enumerable || !("value" in descriptor))
        fail("invalid_data");
      Object.defineProperty(result, key, {
        value: copyJson(descriptor.value, active),
        enumerable: true,
        configurable: true,
        writable: true,
      });
    }
    return result;
  } finally {
    active.delete(value);
  }
}
function deepFreeze<T>(value: T): T {
  if (value !== null && typeof value === "object") {
    for (const item of Object.values(value)) deepFreeze(item);
    Object.freeze(value);
  }
  return value;
}
function statIdentity(value: BigIntStats): string {
  return [
    value.dev,
    value.ino,
    value.uid,
    value.mode,
    value.nlink,
    value.size,
    value.mtimeNs,
    value.ctimeNs,
  ].join(":");
}
function parentIdentity(value: BigIntStats): string {
  return [value.dev, value.ino, value.uid, value.mode].join(":");
}
async function readProtected(
  path: string,
): Promise<{ body: string; fingerprint: string }> {
  let file: FileHandle | undefined;
  try {
    if (typeof process.getuid !== "function") fail("manifest_insecure");
    const owner = BigInt(process.getuid());
    const parent = await lstat(dirname(path), { bigint: true });
    if (
      !parent.isDirectory() ||
      parent.uid !== owner ||
      (parent.mode & 0o7777n) !== 0o700n ||
      (await realpath(path)) !== path
    )
      fail("manifest_insecure");
    file = await open(
      path,
      constants.O_RDONLY | constants.O_NOFOLLOW | constants.O_NONBLOCK,
    );
    const before = await file.stat({ bigint: true });
    if (
      !before.isFile() ||
      before.uid !== owner ||
      (before.mode & 0o7777n) !== 0o600n ||
      before.nlink !== 1n
    )
      fail("manifest_insecure");
    if (before.size > BigInt(MAX_BYTES)) fail("manifest_too_large");
    const buffer = Buffer.alloc(MAX_BYTES + 1);
    let size = 0;
    while (size <= MAX_BYTES) {
      const { bytesRead } = await file.read(
        buffer,
        size,
        buffer.length - size,
        null,
      );
      if (bytesRead === 0) break;
      size += bytesRead;
    }
    if (size > MAX_BYTES) fail("manifest_too_large");
    const after = await file.stat({ bigint: true });
    if (
      statIdentity(before) !== statIdentity(after) ||
      statIdentity(await lstat(path, { bigint: true })) !==
        statIdentity(after) ||
      parentIdentity(parent) !==
        parentIdentity(await lstat(dirname(path), { bigint: true })) ||
      (await realpath(path)) !== path
    )
      fail("manifest_changed");
    const bytes = buffer.subarray(0, size);
    const body = new TextDecoder("utf-8", {
      fatal: true,
      ignoreBOM: true,
    }).decode(bytes);
    return {
      body,
      fingerprint: `${parentIdentity(parent)}:${statIdentity(after)}:${createHash("sha256").update(bytes).digest("hex")}`,
    };
  } catch (error) {
    if (error instanceof OpenClawProductActivationError) throw error;
    return fail("manifest_unavailable");
  } finally {
    try {
      await file?.close();
    } catch {
      fail("manifest_unavailable");
    }
  }
}
